### PR TITLE
Histogram calculation updates

### PR DESCRIPTION
**Pull request by dtorop**
This modernizes the histogram calculation code (the code in common/histogram.c, not to be confused with the UI scope code in libs/histogram.c. It removes the hand-written SSE code, but the generated code should be equivalent in speed. It removes approx. 189 LOC.

Changes:

- Remove the code to generate a histogram of a floating point raw images. This appears to be unused anywhere in the current darktable codebase.
- The exposure deflicker code previously bypassed the standard histogram call `dt_histogram_helper()`. It now uses that call to generate a histogram of the raw image data (which will be 16 bit unsigned int).
- The histogram data buffer is now 64-bit aligned. As before the caller to `dt_histogram_helper()` must free any returned buffer, but now it must use `dt_free_align()`
- As mentioned, all the SSE code is gone. But I used it as a model for the revised code (which relies heavily on the `for_each_channel()`), hence the generated assembly should look eerily similar in many cases.
- ~~Clamping of bin numbers is now always done by converting them to integers first. Previously they were sometimes clamped as floats. This shouldn't make any difference in the results, but is more "correct".~~ Wrong! Clamping negative floats as unsigned integers is not advisable.
- The `mul` parameter was removed from `dt_dev_histogram_collection_params_t`. It appears to have been intended for scaling the binning to something other than the maximum # of bins, but no current code used this feature.
- A `buf_size` parameter was added to `dt_dev_histogram_stats_t`. This allows the histogram code to calculate if the # of bins has changed, and if so allocate a larger buffer if necessary. (Changing of the buffer size is only used by the levels iop. It might be a bit more slick to put this tracking/reallocating responsibility on the levels iop.)
- `dt_histogram_max_helper()` as a separate call was eliminated. Pass the `histogram_max` array in to the main `dt_histogram_helper()` function instead, or `NULL` if no maximum calculation is needed (which is the case for exposure deflicker).
- The prior code ignored the first bin when calculating max values for each channel. This makes sense for the `R`, `G`, `B`, `L`, and `C` channels to avoid scaling to underexposed values. But not so much for the `a`, `b` and `h` values, so bin 0 for these is counted.
- Rather than using hand-rolled per-thread buffers and explicitly written reduction code when using OpenMP threads, the code relies on OpenMP array reductions. This simplifies the code, and there appears to be no speed difference.
- `compensate_middle_grey` is now a boolean (it was an int)

Rationale for this work:
- Compilers are increasingly capable, and it doesn't make sense in many cases to handcode SIMD instructions.
- OpenMP 4.5 allows for array reductions, which further reduces code size.
- The prior histogram calculation code, when perused, looks like something magic. Though this PR's code is much less magic, it is more succinct, and hence should be easier to reason about.
- There was some idea that the crashes on Windows are caused by unaligned access to histogram data. Though I don't see any huge signs of problems here, this code does align the histogram data, which might help.
- There has been so much revision in the use of histogram code over the years that it makes sense to perform an overhaul. This allows for discovering unused code, etc. As the actual uses of histograms in darktable has become clear, it makes sense to code the histogram calculation around how the histograms are actually used, rather than coding something slightly more generalized.
- Prior code often had OpenMP and SSE execution paths for the same operation. This PR only has one execution path for each operation.

### CODE REVIEW

**Comment by TurboGit** [comment]
@dtorop  : The CI failed with:

```
 /Users/runner/work/darktable/darktable/src/src/common/histogram.c:175:29: fatal error: variable 'bins_total' must have explicitly specified data sharing attributes
  reduction(+:working_hist[:bins_total])                                  \
                            ^~~~~~~~~~
/Users/runner/work/darktable/darktable/src/src/common/histogram.c:173:34: note: explicit data sharing attribute requested here
#pragma omp parallel for default(none)                                    \
                                 ^
1 error generated.
```

**Comment by dtorop** [comment]
> @dtorop : The CI failed with:
...
> 1 error generated.

Drat! It's a LLVM thing, but it's fixable. There's also an error with converting negative values to unsigned int, so a couple fixes on the way.

**Comment by dtorop** [comment]
Last commit should fix LLVM compilation, as well as a bug on my part which produced bad results with negative input, and generally polishes things.

A remaining worry: Does the OpenMP array reduction used in this PR for the binned data use the stack for storing per-thread data? This isn't a worry for most histograms (256 bins is 4kb/thread, and stack is 8MB). But levels can ask for 16384 bins, and exposure's deflicker asks for 65536 bins.

**Comment by dtorop** [comment]
Looks like CI failed due to network problem. I force-pushed the last commit to convince it to run again.

File: src/common/histogram.c [hunk 1]
```diff
@@ -29,332 +29,163 @@
...
-                                           const dt_iop_order_iccprofile_info_t *const profile_info)
-{
-  const dt_histogram_roi_t *roi = histogram_params->roi;
-  const float *input = (float *)pixel + roi->width * j + roi->crop_x;
-  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++, input++)
-  {
-    histogram_helper_cs_RAW_helper_process_pixel_float(histogram_params, input, histogram);
-  }
+  DT_ALIGNED_PIXEL size_t bin[4];
+  for_each_channel(k,aligned(vals,bin:16))
```
Comment (TurboGit) [inline src/common/histogram.c:42]: please enclose with `{ }` the loop instructions. With the comment in between the first line of code it is a source of errors for maintenance.

File: src/common/histogram.c [hunk 2]
```diff
@@ -29,332 +29,163 @@
...
-{
-  const dt_histogram_roi_t *roi = histogram_params->roi;
+  const dt_histogram_roi_t *roi = params->roi;
   uint16_t *in = (uint16_t *)pixel + roi->width * j + roi->crop_x;
+  const size_t max_bin = params->bins_count - 1;
 
-  // process pixels
-  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++, in++)
-    histogram_helper_cs_RAW_helper_process_pixel_uint16(histogram_params, in, histogram);
+  for(int i = 0; i < roi->width - roi->crop_width - roi->crop_x; i++)
```
Comment (TurboGit) [inline src/common/histogram.c:61]: please enclose with `{ }` the loop instructions. With the comment in between the first line of code it is a source of errors for maintenance.

**Review by TurboGit — APPROVED** [review]
All good, will merge and will fix the code as per comment.

**Comment by dtorop** [comment]
@TurboGit: Thank you! And thanks for {} cleanups...

I'll be curious if this helps Windows instability. At least it should be slightly less scary to debug that now.

**Comment by dtorop** [comment]
@TurboGit: Belated release notes:

"Modernize the histogram calculation code. Remove SSE code (which provides no speed-ups), but use it as a model for the optimized code using recent OpenMP features. Remove various unused bits of code, and provide a consistent internal API. In certain cases this code will produce marginally more accurate results. In some cases the new code uses substantially less memory."

Again, I can make this more succinct!

**Comment by TurboGit** [comment]
Thanks!

**Pull request merged**